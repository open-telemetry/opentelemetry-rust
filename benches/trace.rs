use criterion::{criterion_group, criterion_main, Criterion};
use opentelemetry::{
    sdk::trace as sdktrace,
    trace::{Span, Tracer, TracerProvider},
    Key,
};

fn criterion_benchmark(c: &mut Criterion) {
    trace_benchmark_group(c, "start-end-span", |tracer| tracer.start("foo").end());

    trace_benchmark_group(c, "start-end-span-4-attrs", |tracer| {
        let span = tracer.start("foo");
        span.set_attribute(Key::new("key1").bool(false));
        span.set_attribute(Key::new("key2").string("hello"));
        span.set_attribute(Key::new("key4").f64(123.456));
        span.end();
    });

    trace_benchmark_group(c, "start-end-span-8-attrs", |tracer| {
        let span = tracer.start("foo");
        span.set_attribute(Key::new("key1").bool(false));
        span.set_attribute(Key::new("key2").string("hello"));
        span.set_attribute(Key::new("key4").f64(123.456));
        span.set_attribute(Key::new("key11").bool(false));
        span.set_attribute(Key::new("key12").string("hello"));
        span.set_attribute(Key::new("key14").f64(123.456));
        span.end();
    });

    trace_benchmark_group(c, "start-end-span-all-attr-types", |tracer| {
        let span = tracer.start("foo");
        span.set_attribute(Key::new("key1").bool(false));
        span.set_attribute(Key::new("key2").string("hello"));
        span.set_attribute(Key::new("key3").i64(123));
        span.set_attribute(Key::new("key5").f64(123.456));
        span.end();
    });

    trace_benchmark_group(c, "start-end-span-all-attr-types-2x", |tracer| {
        let span = tracer.start("foo");
        span.set_attribute(Key::new("key1").bool(false));
        span.set_attribute(Key::new("key2").string("hello"));
        span.set_attribute(Key::new("key3").i64(123));
        span.set_attribute(Key::new("key5").f64(123.456));
        span.set_attribute(Key::new("key11").bool(false));
        span.set_attribute(Key::new("key12").string("hello"));
        span.set_attribute(Key::new("key13").i64(123));
        span.set_attribute(Key::new("key15").f64(123.456));
        span.end();
    });
}

fn trace_benchmark_group<F: Fn(&sdktrace::Tracer)>(c: &mut Criterion, name: &str, f: F) {
    let mut group = c.benchmark_group(name);

    group.bench_function("always-sample", |b| {
        let always_sample = sdktrace::TracerProvider::builder()
            .with_config(sdktrace::Config {
                default_sampler: Box::new(sdktrace::Sampler::AlwaysOn),
                ..Default::default()
            })
            .build()
            .get_tracer("always-sample", None);

        b.iter(|| f(&always_sample));
    });

    group.bench_function("never-sample", |b| {
        let never_sample = sdktrace::TracerProvider::builder()
            .with_config(sdktrace::Config {
                default_sampler: Box::new(sdktrace::Sampler::AlwaysOff),
                ..Default::default()
            })
            .build()
            .get_tracer("never-sample", None);
        b.iter(|| f(&never_sample));
    });

    group.finish();
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
