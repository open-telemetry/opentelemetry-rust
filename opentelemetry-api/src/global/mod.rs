//! Utilities for working with global telemetry primitives
//!
//! ## Global Trace API
//!
//! The global trace API **provides applications access to their configured
//! [`TracerProvider`] instance from anywhere in the codebase**. This allows
//! applications to be less coupled to the specific Open Telemetry SDK while not
//! manually passing references to each part of the code that needs to create
//! [`Span`]s. Additionally, **3rd party middleware** or **library code** can be
//! written against this generic API and not constrain users to a specific
//! implementation choice.
//!
//! ### Usage in Applications
//!
//! Applications configure their tracer either by installing a trace pipeline,
//! or calling [`set_tracer_provider`].
//!
//! ```
//! # #[cfg(feature="trace")]
//! # {
//! use opentelemetry_api::trace::{Tracer, noop::NoopTracerProvider};
//! use opentelemetry_api::global;
//!
//! fn init_tracer() {
//!     // Swap this no-op provider for your tracing service of choice (jaeger, zipkin, etc)
//!     let provider = NoopTracerProvider::new();
//!
//!     // Configure the global `TracerProvider` singleton when your app starts
//!     // (there is a no-op default if this is not set by your application)
//!     let _ = global::set_tracer_provider(provider);
//! }
//!
//! fn do_something_tracked() {
//!     // Then you can get a named tracer instance anywhere in your codebase.
//!     let tracer = global::tracer("my-component");
//!
//!     tracer.in_span("doing_work", |cx| {
//!         // Traced app logic here...
//!     });
//! }
//!
//! // in main or other app start
//! init_tracer();
//! do_something_tracked();
//! # }
//! ```
//!
//! ### Usage in Libraries
//!
//! ```
//! # #[cfg(feature="trace")]
//! # {
//! use opentelemetry_api::trace::{Tracer, TracerProvider};
//! use opentelemetry_api::global;
//!
//! pub fn my_traced_library_function() {
//!     // End users of your library will configure their global tracer provider
//!     // so you can use the global tracer without any setup
//!     let tracer = global::tracer_provider().versioned_tracer(
//!         "my-library-name",
//!         Some(env!("CARGO_PKG_VERSION")),
//!         None,
//!     );
//!
//!     tracer.in_span("doing_library_work", |cx| {
//!         // Traced library logic here...
//!     });
//! }
//! # }
//! ```
//!
//! [`TracerProvider`]: crate::trace::TracerProvider
//! [`Span`]: crate::trace::Span
//!
//! ## Global Metrics API
//!
//! The global metrics API **provides applications access to their configured
//! [`MeterProvider`] instance from anywhere in the codebase**. This allows
//! applications to be less coupled to the specific Open Telemetry SDK while not
//! manually passing references to each part of the code that needs to create
//! metric instruments. Additionally, **3rd party middleware** or **library code** can be
//! written against this generic API and not constrain users to a specific
//! implementation choice.
//!
//! ### Usage in Applications
//!
//! Applications configure their meter either by installing a metrics pipeline,
//! or calling [`set_meter_provider`].
//!
//! ```
//! # #[cfg(feature="metrics")]
//! # {
//! use opentelemetry_api::metrics::{Meter, noop::NoopMeterProvider};
//! use opentelemetry_api::{global, Context, KeyValue};
//!
//! fn init_meter() {
//!     let provider = NoopMeterProvider::new();
//!
//!     // Configure the global `MeterProvider` singleton when your app starts
//!     // (there is a no-op default if this is not set by your application)
//!     global::set_meter_provider(provider)
//! }
//!
//! fn do_something_instrumented() {
//!     // Then you can get a named tracer instance anywhere in your codebase.
//!     let meter = global::meter("my-component");
//!     let counter = meter.u64_counter("my_counter").init();
//!     let cx = Context::current();
//!
//!     // record metrics
//!     counter.add(&cx, 1, &[KeyValue::new("mykey", "myvalue")]);
//! }
//!
//! // in main or other app start
//! init_meter();
//! do_something_instrumented();
//! # }
//! ```
//!
//! ### Usage in Libraries
//!
//! ```
//! # #[cfg(feature="metrics")]
//! # {
//! use opentelemetry_api::{global, Context, KeyValue};
//!
//! pub fn my_traced_library_function() {
//!     // End users of your library will configure their global meter provider
//!     // so you can use the global meter without any setup
//!     let tracer = global::meter("my-library-name");
//!     let counter = tracer.u64_counter("my_counter").init();
//!     let cx = Context::current();
//!
//!     // record metrics
//!     counter.add(&cx, 1, &[KeyValue::new("mykey", "myvalue")]);
//! }
//! # }
//! ```
//!
//! [`MeterProvider`]: crate::metrics::MeterProvider
//! [`set_meter_provider`]: crate::global::set_meter_provider

mod error_handler;
#[cfg(feature = "metrics")]
mod metrics;
#[cfg(feature = "trace")]
mod propagation;
#[cfg(feature = "trace")]
mod trace;

pub use error_handler::{handle_error, set_error_handler, Error};
#[cfg(feature = "metrics")]
#[cfg_attr(docsrs, doc(cfg(feature = "metrics")))]
pub use metrics::{
    meter, meter_provider, meter_with_version, set_meter_provider, GlobalMeterProvider,
};
#[cfg(feature = "trace")]
#[cfg_attr(docsrs, doc(cfg(feature = "trace")))]
pub use propagation::{get_text_map_propagator, set_text_map_propagator};
#[cfg(feature = "trace")]
#[cfg_attr(docsrs, doc(cfg(feature = "trace")))]
pub use trace::{
    set_tracer_provider, shutdown_tracer_provider, tracer, tracer_provider, BoxedSpan, BoxedTracer,
    GlobalTracerProvider, ObjectSafeTracer, ObjectSafeTracerProvider,
};
