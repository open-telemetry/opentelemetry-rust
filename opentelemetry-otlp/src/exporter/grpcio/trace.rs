use core::fmt;
use std::{collections::HashMap, time::Duration};

use futures_core::future::BoxFuture;
use grpcio::CallOption;
use opentelemetry_api::trace::TraceError;
use opentelemetry_proto::grpcio::{
    trace_service::ExportTraceServiceRequest, trace_service_grpc::TraceServiceClient,
};
use opentelemetry_sdk::export::trace::{ExportResult, SpanData, SpanExporter};

use grpcio::MetadataBuilder;

pub(crate) struct GrpcioTraceClient {
    client: Option<TraceServiceClient>,
    timeout: Duration,
    headers: HashMap<String, String>,
}

impl fmt::Debug for GrpcioTraceClient {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("GrpcioTracesClient")
    }
}

impl GrpcioTraceClient {
    /// Create a new trace client
    pub(crate) fn new(
        client: TraceServiceClient,
        timeout: Duration,
        headers: HashMap<String, String>,
    ) -> Self {
        GrpcioTraceClient {
            client: Some(client),
            timeout,
            headers,
        }
    }
}

impl SpanExporter for GrpcioTraceClient {
    fn export(&mut self, batch: Vec<SpanData>) -> BoxFuture<'static, ExportResult> {
        let client = match &self.client {
            Some(client) => client.clone(),
            None => {
                return Box::pin(std::future::ready(Err(TraceError::Other(
                    "exporter is already shut down".into(),
                ))))
            }
        };

        let request = ExportTraceServiceRequest {
            resource_spans: protobuf::RepeatedField::from_vec(
                batch.into_iter().map(Into::into).collect(),
            ),
            ..Default::default()
        };

        let mut call_options = CallOption::default().timeout(self.timeout);

        if !self.headers.is_empty() {
            let mut metadata_builder: MetadataBuilder = MetadataBuilder::new();

            for (key, value) in self.headers.iter() {
                let _ = metadata_builder.add_str(key.as_str(), value.as_str());
            }

            call_options = call_options.headers(metadata_builder.build());
        }

        Box::pin(async move {
            let receiver = client
                .export_async_opt(&request, call_options)
                .map_err(crate::Error::from)?;

            receiver.await.map_err(crate::Error::from)?;

            Ok(())
        })
    }

    fn shutdown(&mut self) {
        let _ = self.client.take();
    }
}
