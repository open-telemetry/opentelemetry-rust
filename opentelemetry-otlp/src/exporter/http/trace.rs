use std::sync::Arc;

use futures_core::future::BoxFuture;
use http::{header::CONTENT_TYPE, Method};
use opentelemetry::trace::TraceError;
use opentelemetry_sdk::export::trace::{ExportResult, SpanData, SpanExporter};

use super::OtlpHttpClient;

impl SpanExporter for OtlpHttpClient {
    fn export(&mut self, batch: Vec<SpanData>) -> BoxFuture<'static, ExportResult> {
        let client = match self
            .client
            .lock()
            .map_err(|e| TraceError::Other(e.to_string().into()))
            .and_then(|g| match &*g {
                Some(client) => Ok(Arc::clone(client)),
                _ => Err(TraceError::Other("exporter is already shut down".into())),
            }) {
            Ok(client) => client,
            Err(err) => return Box::pin(std::future::ready(Err(err))),
        };

        let (body, content_type) = match self.build_trace_export_body(batch, &self.resource) {
            Ok(body) => body,
            Err(e) => return Box::pin(std::future::ready(Err(e))),
        };

        let mut request = match http::Request::builder()
            .method(Method::POST)
            .uri(&self.collector_endpoint)
            .header(CONTENT_TYPE, content_type)
            .body(body)
        {
            Ok(req) => req,
            Err(e) => {
                return Box::pin(std::future::ready(Err(crate::Error::RequestFailed(
                    Box::new(e),
                )
                .into())))
            }
        };

        for (k, v) in &self.headers {
            request.headers_mut().insert(k.clone(), v.clone());
        }

        Box::pin(async move {
            let request_uri = request.uri().to_string();
            let response = client.send(request).await?;

            if !response.status().is_success() {
                let error = format!(
                    "OpenTelemetry trace export failed. Url: {}, Status Code: {}, Response: {:?}",
                    response.status().as_u16(),
                    request_uri,
                    response.body()
                );
                return Err(TraceError::Other(error.into()));
            }

            Ok(())
        })
    }

    fn shutdown(&mut self) {
        let _ = self.client.lock().map(|mut c| c.take());
    }

    fn set_resource(&mut self, resource: &opentelemetry_sdk::Resource) {
        self.resource = resource.into();
    }
}
