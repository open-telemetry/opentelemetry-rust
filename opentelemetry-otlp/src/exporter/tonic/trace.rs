use core::fmt;

use futures_core::future::BoxFuture;
use opentelemetry::trace::TraceError;
use opentelemetry_proto::tonic::collector::trace::v1::{
    trace_service_client::TraceServiceClient, ExportTraceServiceRequest,
};
use opentelemetry_proto::tonic::trace::v1::ResourceSpans;
use opentelemetry_sdk::export::trace::{ExportResult, SpanData, SpanExporter};
use tonic::{codegen::CompressionEncoding, service::Interceptor, transport::Channel, Request};

use super::BoxInterceptor;

pub(crate) struct TonicTracesClient {
    inner: Option<ClientInner>,
    #[allow(dead_code)]
    // <allow dead> would be removed once we support set_resource for metrics.
    resource: opentelemetry_proto::transform::common::tonic::ResourceAttributesWithSchema,
}

struct ClientInner {
    client: TraceServiceClient<Channel>,
    interceptor: BoxInterceptor,
}

impl fmt::Debug for TonicTracesClient {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("TonicTracesClient")
    }
}

impl TonicTracesClient {
    pub(super) fn new(
        channel: Channel,
        interceptor: BoxInterceptor,
        compression: Option<CompressionEncoding>,
    ) -> Self {
        let mut client = TraceServiceClient::new(channel);
        if let Some(compression) = compression {
            client = client
                .send_compressed(compression)
                .accept_compressed(compression);
        }

        TonicTracesClient {
            inner: Some(ClientInner {
                client,
                interceptor,
            }),
            resource: Default::default(),
        }
    }
}

impl SpanExporter for TonicTracesClient {
    fn export(&mut self, batch: Vec<SpanData>) -> BoxFuture<'static, ExportResult> {
        let (mut client, metadata, extensions) = match &mut self.inner {
            Some(inner) => {
                let (m, e, _) = match inner.interceptor.call(Request::new(())) {
                    Ok(res) => res.into_parts(),
                    Err(e) => {
                        return Box::pin(std::future::ready(Err(TraceError::Other(Box::new(e)))))
                    }
                };
                (inner.client.clone(), m, e)
            }
            None => {
                return Box::pin(std::future::ready(Err(TraceError::Other(
                    "exporter is already shut down".into(),
                ))))
            }
        };

        // TODO: Avoid cloning here.
        let resource_spans = {
            batch
                .into_iter()
                .map(|log_data| ResourceSpans::new(log_data, &self.resource))
                .collect()
        };

        Box::pin(async move {
            client
                .export(Request::from_parts(
                    metadata,
                    extensions,
                    ExportTraceServiceRequest { resource_spans },
                ))
                .await
                .map_err(crate::Error::from)?;

            Ok(())
        })
    }

    fn shutdown(&mut self) {
        let _ = self.inner.take();
    }

    fn set_resource(&mut self, resource: &opentelemetry_sdk::Resource) {
        self.resource = resource.into();
    }
}
