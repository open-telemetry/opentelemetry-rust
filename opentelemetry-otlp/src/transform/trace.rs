#[cfg(any(feature = "http-proto", feature = "http-json", feature = "grpc-tonic"))]
/// Tonic-specific transformation utilities for traces.
pub mod tonic {
    use opentelemetry_proto::tonic::resource::v1::Resource;
    use opentelemetry_proto::tonic::trace::v1::{span, status, ResourceSpans, ScopeSpans, Span, Status};
    use crate::transform::common::{
        to_nanos,
        tonic::{Attributes, ResourceAttributesWithSchema, instrumentation_scope_from_scope_and_target, instrumentation_scope_from_scope_ref_and_target},
    };
    use opentelemetry::trace;
    use opentelemetry::trace::{Link, SpanId, SpanKind};
    use opentelemetry_sdk::trace::SpanData;
    use std::collections::HashMap;

    /// Converts SDK span kind to protobuf span kind.
    pub fn span_kind_to_proto_span_kind(span_kind: SpanKind) -> span::SpanKind {
        match span_kind {
            SpanKind::Client => span::SpanKind::Client,
            SpanKind::Consumer => span::SpanKind::Consumer,
            SpanKind::Internal => span::SpanKind::Internal,
            SpanKind::Producer => span::SpanKind::Producer,
            SpanKind::Server => span::SpanKind::Server,
        }
    }

    /// Converts SDK trace status to protobuf status code.
    pub fn trace_status_to_proto_status_code(status: &trace::Status) -> status::StatusCode {
        match status {
            trace::Status::Ok => status::StatusCode::Ok,
            trace::Status::Unset => status::StatusCode::Unset,
            trace::Status::Error { .. } => status::StatusCode::Error,
        }
    }

    /// Converts SDK link to protobuf link.
    pub fn link_to_proto_link(link: Link) -> span::Link {
        span::Link {
            trace_id: link.span_context.trace_id().to_bytes().to_vec(),
            span_id: link.span_context.span_id().to_bytes().to_vec(),
            trace_state: link.span_context.trace_state().header(),
            attributes: Attributes::from(link.attributes).0,
            dropped_attributes_count: link.dropped_attributes_count,
            flags: link.span_context.trace_flags().to_u8() as u32,
        }
    }

    /// Converts SDK span data to protobuf span.
    pub fn span_data_to_proto_span(source_span: opentelemetry_sdk::trace::SpanData) -> Span {
        let span_kind: span::SpanKind = span_kind_to_proto_span_kind(source_span.span_kind);
        Span {
            trace_id: source_span.span_context.trace_id().to_bytes().to_vec(),
            span_id: source_span.span_context.span_id().to_bytes().to_vec(),
            trace_state: source_span.span_context.trace_state().header(),
            parent_span_id: {
                if source_span.parent_span_id != SpanId::INVALID {
                    source_span.parent_span_id.to_bytes().to_vec()
                } else {
                    vec![]
                }
            },
            flags: source_span.span_context.trace_flags().to_u8() as u32,
            name: source_span.name.into_owned(),
            kind: span_kind as i32,
            start_time_unix_nano: to_nanos(source_span.start_time),
            end_time_unix_nano: to_nanos(source_span.end_time),
            dropped_attributes_count: source_span.dropped_attributes_count,
            attributes: Attributes::from(source_span.attributes).0,
            dropped_events_count: source_span.events.dropped_count,
            events: source_span
                .events
                .into_iter()
                .map(|event| span::Event {
                    time_unix_nano: to_nanos(event.timestamp),
                    name: event.name.into(),
                    attributes: Attributes::from(event.attributes).0,
                    dropped_attributes_count: event.dropped_attributes_count,
                })
                .collect(),
            dropped_links_count: source_span.links.dropped_count,
            links: source_span.links.into_iter().map(link_to_proto_link).collect(),
            status: Some(Status {
                code: trace_status_to_proto_status_code(&source_span.status).into(),
                message: match source_span.status {
                    trace::Status::Error { description } => description.to_string(),
                    _ => Default::default(),
                },
            }),
        }
    }

    /// Creates a new resource spans from span data and resource.
    pub fn new_resource_spans(source_span: SpanData, resource: &ResourceAttributesWithSchema) -> ResourceSpans {
        ResourceSpans {
            resource: Some(Resource {
                attributes: resource.attributes.0.clone(),
                dropped_attributes_count: 0,
                entity_refs: vec![],
            }),
            schema_url: resource.schema_url.clone().unwrap_or_default(),
            scope_spans: vec![ScopeSpans {
                schema_url: source_span
                    .instrumentation_scope
                    .schema_url()
                    .map(ToOwned::to_owned)
                    .unwrap_or_default(),
                scope: Some(instrumentation_scope_from_scope_and_target(source_span.instrumentation_scope.clone(), None)),
                spans: vec![span_data_to_proto_span(source_span)],
            }],
        }
    }

    /// Groups spans by resource and instrumentation scope.
    pub fn group_spans_by_resource_and_scope(
        spans: Vec<SpanData>,
        resource: &ResourceAttributesWithSchema,
    ) -> Vec<ResourceSpans> {
        // Group spans by their instrumentation scope
        let scope_map = spans.iter().fold(
            HashMap::new(),
            |mut scope_map: HashMap<&opentelemetry::InstrumentationScope, Vec<&SpanData>>, span| {
                let instrumentation = &span.instrumentation_scope;
                scope_map.entry(instrumentation).or_default().push(span);
                scope_map
            },
        );

        // Convert the grouped spans into ScopeSpans
        let scope_spans = scope_map
            .into_iter()
            .map(|(instrumentation, span_records)| ScopeSpans {
                scope: Some(instrumentation_scope_from_scope_ref_and_target(instrumentation, None)),
                schema_url: resource.schema_url.clone().unwrap_or_default(),
                spans: span_records
                    .into_iter()
                    .map(|span_data| span_data_to_proto_span(span_data.clone()))
                    .collect(),
            })
            .collect();

        // Wrap ScopeSpans into a single ResourceSpans
        vec![ResourceSpans {
            resource: Some(Resource {
                attributes: resource.attributes.0.clone(),
                dropped_attributes_count: 0,
                entity_refs: vec![],
            }),
            scope_spans,
            schema_url: resource.schema_url.clone().unwrap_or_default(),
        }]
    }

} // End of tonic module

#[cfg(test)]
mod tests {
    use opentelemetry_proto::tonic::common::v1::any_value::Value;
    use crate::transform::common::tonic::ResourceAttributesWithSchema;
    use opentelemetry::time::now;
    use opentelemetry::trace::{
        SpanContext, SpanId, SpanKind, Status, TraceFlags, TraceId, TraceState,
    };
    use opentelemetry::InstrumentationScope;
    use opentelemetry::KeyValue;
    use opentelemetry_sdk::resource::Resource;
    use opentelemetry_sdk::trace::SpanData;
    use opentelemetry_sdk::trace::{SpanEvents, SpanLinks};
    use std::borrow::Cow;
    use std::time::Duration;

    fn create_test_span_data(instrumentation_name: &'static str) -> SpanData {
        let span_context = SpanContext::new(
            TraceId::from_u128(123),
            SpanId::from_u64(456),
            TraceFlags::default(),
            false,
            TraceState::default(),
        );

        SpanData {
            span_context,
            parent_span_id: SpanId::from_u64(0),
            span_kind: SpanKind::Internal,
            name: Cow::Borrowed("test_span"),
            start_time: now(),
            end_time: now() + Duration::from_secs(1),
            attributes: vec![KeyValue::new("key", "value")],
            dropped_attributes_count: 0,
            events: SpanEvents::default(),
            links: SpanLinks::default(),
            status: Status::Unset,
            instrumentation_scope: InstrumentationScope::builder(instrumentation_name).build(),
        }
    }

    #[test]
    fn test_group_spans_by_resource_and_scope_single_scope() {
        let resource = Resource::builder_empty()
            .with_attribute(KeyValue::new("resource_key", "resource_value"))
            .build();
        let span_data = create_test_span_data("lib1");

        let spans = vec![span_data.clone()];
        let resource: ResourceAttributesWithSchema = (&resource).into(); // Convert Resource to ResourceAttributesWithSchema

        let grouped_spans =
            super::tonic::group_spans_by_resource_and_scope(spans, &resource);

        assert_eq!(grouped_spans.len(), 1);

        let resource_spans = &grouped_spans[0];
        assert_eq!(
            resource_spans.resource.as_ref().unwrap().attributes.len(),
            1
        );
        assert_eq!(
            resource_spans.resource.as_ref().unwrap().attributes[0].key,
            "resource_key"
        );
        assert_eq!(
            resource_spans.resource.as_ref().unwrap().attributes[0]
                .value
                .clone()
                .unwrap()
                .value
                .unwrap(),
            Value::StringValue("resource_value".to_string())
        );

        let scope_spans = &resource_spans.scope_spans;
        assert_eq!(scope_spans.len(), 1);

        let scope_span = &scope_spans[0];
        assert_eq!(scope_span.scope.as_ref().unwrap().name, "lib1");
        assert_eq!(scope_span.spans.len(), 1);

        assert_eq!(
            scope_span.spans[0].trace_id,
            span_data.span_context.trace_id().to_bytes().to_vec()
        );
    }

    #[test]
    fn test_group_spans_by_resource_and_scope_multiple_scopes() {
        let resource = Resource::builder_empty()
            .with_attribute(KeyValue::new("resource_key", "resource_value"))
            .build();
        let span_data1 = create_test_span_data("lib1");
        let span_data2 = create_test_span_data("lib1");
        let span_data3 = create_test_span_data("lib2");

        let spans = vec![span_data1.clone(), span_data2.clone(), span_data3.clone()];
        let resource: ResourceAttributesWithSchema = (&resource).into(); // Convert Resource to ResourceAttributesWithSchema

        let grouped_spans =
            super::tonic::group_spans_by_resource_and_scope(spans, &resource);

        assert_eq!(grouped_spans.len(), 1);

        let resource_spans = &grouped_spans[0];
        assert_eq!(
            resource_spans.resource.as_ref().unwrap().attributes.len(),
            1
        );
        assert_eq!(
            resource_spans.resource.as_ref().unwrap().attributes[0].key,
            "resource_key"
        );
        assert_eq!(
            resource_spans.resource.as_ref().unwrap().attributes[0]
                .value
                .clone()
                .unwrap()
                .value
                .unwrap(),
            Value::StringValue("resource_value".to_string())
        );

        let scope_spans = &resource_spans.scope_spans;
        assert_eq!(scope_spans.len(), 2);

        // Check the scope spans for both lib1 and lib2
        let mut lib1_scope_span = None;
        let mut lib2_scope_span = None;

        for scope_span in scope_spans {
            match scope_span.scope.as_ref().unwrap().name.as_str() {
                "lib1" => lib1_scope_span = Some(scope_span),
                "lib2" => lib2_scope_span = Some(scope_span),
                _ => {}
            }
        }

        let lib1_scope_span = lib1_scope_span.expect("lib1 scope span not found");
        let lib2_scope_span = lib2_scope_span.expect("lib2 scope span not found");

        assert_eq!(lib1_scope_span.scope.as_ref().unwrap().name, "lib1");
        assert_eq!(lib2_scope_span.scope.as_ref().unwrap().name, "lib2");

        assert_eq!(lib1_scope_span.spans.len(), 2);
        assert_eq!(lib2_scope_span.spans.len(), 1);

        assert_eq!(
            lib1_scope_span.spans[0].trace_id,
            span_data1.span_context.trace_id().to_bytes().to_vec()
        );
        assert_eq!(
            lib1_scope_span.spans[1].trace_id,
            span_data2.span_context.trace_id().to_bytes().to_vec()
        );
        assert_eq!(
            lib2_scope_span.spans[0].trace_id,
            span_data3.span_context.trace_id().to_bytes().to_vec()
        );
    }
}
