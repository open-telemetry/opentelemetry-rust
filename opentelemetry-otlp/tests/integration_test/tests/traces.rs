use integration_test_runner::asserter::{read_spans_from_json, TraceAsserter};
use opentelemetry::global;
use opentelemetry::global::shutdown_tracer_provider;
use opentelemetry::trace::TraceError;
use opentelemetry::{
    trace::{TraceContextExt, Tracer},
    Key, KeyValue,
};
use opentelemetry_sdk::{runtime, trace as sdktrace, Resource};
use std::error::Error;
use std::fs::File;
use std::os::unix::fs::MetadataExt;

fn init_tracer() -> Result<sdktrace::Tracer, TraceError> {
    opentelemetry_otlp::new_pipeline()
        .tracing()
        .with_exporter(opentelemetry_otlp::new_exporter().tonic())
        .with_trace_config(
            sdktrace::config().with_resource(Resource::new(vec![KeyValue::new(
                opentelemetry_semantic_conventions::resource::SERVICE_NAME,
                "basic-otlp-tracing-example",
            )])),
        )
        .install_batch(runtime::Tokio)
}

const LEMONS_KEY: Key = Key::from_static_str("lemons");
const ANOTHER_KEY: Key = Key::from_static_str("ex.com/another");

pub async fn traces() -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
    // By binding the result to an unused variable, the lifetime of the variable
    // matches the containing block, reporting traces and metrics during the whole
    // execution.
    let _ = init_tracer()?;

    let tracer = global::tracer("ex.com/basic");

    tracer.in_span("operation", |cx| {
        let span = cx.span();
        span.add_event(
            "Nice operation!".to_string(),
            vec![Key::new("bogons").i64(100)],
        );
        span.set_attribute(KeyValue::new(ANOTHER_KEY, "yes"));

        tracer.in_span("Sub operation...", |cx| {
            let span = cx.span();
            span.set_attribute(KeyValue::new(LEMONS_KEY, "five"));

            span.add_event("Sub span event", vec![]);
        });
    });

    shutdown_tracer_provider();

    Ok(())
}

pub fn assert_traces_results(result: &str, _: &str) {
    // let left = read_spans_from_json(File::open(expected).unwrap());
    // let right = read_spans_from_json(File::open(result).unwrap());

    // TraceAsserter::new(left, right).assert();
    //
    // we cannot read result json file because the timestamp was represents as string instead of u64.
    // need to fix it on json file exporter

    assert!(File::open(result).unwrap().metadata().unwrap().size() > 0)
}

#[test]
#[should_panic(expected = "left: \"Sub operation...\"")] // we swap the parent spans with child spans in failed_traces.json
pub fn test_assert_span_eq_failure() {
    let left = read_spans_from_json(File::open("./expected/traces.json").unwrap());
    let right = read_spans_from_json(File::open("./expected/failed_traces.json").unwrap());

    TraceAsserter::new(right, left).assert();
}

#[test]
pub fn test_assert_span_eq() {
    let spans = read_spans_from_json(File::open("./expected/traces.json").unwrap());

    TraceAsserter::new(spans.clone(), spans).assert();
}
