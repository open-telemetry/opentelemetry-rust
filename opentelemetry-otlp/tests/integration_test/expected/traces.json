{
  "resourceSpans": [
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "basic-otlp-tracing-example"
            }
          }
        ]
      },
      "scopeSpans": [
        {
          "scope": {
            "name": "ex.com/basic"
          },
          "spans": [
            {
              "traceId": "9b458af7378cba65253d7042d34fc72e",
              "spanId": "cd7cf7bf939930b7",
              "parentSpanId": "d58cf2d702a061e0",
              "name": "Sub operation...",
              "kind": 1,
              "startTimeUnixNano": "1703985537070566698",
              "endTimeUnixNano": "1703985537070572718",
              "attributes": [
                {
                  "key": "lemons",
                  "value": {
                    "stringValue": "five"
                  }
                }
              ],
              "events": [
                {
                  "timeUnixNano": "1703985537070567697",
                  "name": "Sub span event"
                }
              ],
              "status": {}
            }
          ]
        }
      ]
    },
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "basic-otlp-tracing-example"
            }
          }
        ]
      },
      "scopeSpans": [
        {
          "scope": {
            "name": "ex.com/basic"
          },
          "spans": [
            {
              "traceId": "9b458af7378cba65253d7042d34fc72e",
              "spanId": "d58cf2d702a061e0",
              "parentSpanId": "",
              "name": "operation",
              "kind": 1,
              "startTimeUnixNano": "1703985537070558635",
              "endTimeUnixNano": "1703985537070580454",
              "attributes": [
                {
                  "key": "ex.com/another",
                  "value": {
                    "stringValue": "yes"
                  }
                }
              ],
              "events": [
                {
                  "timeUnixNano": "1703985537070563326",
                  "name": "Nice operation!",
                  "attributes": [
                    {
                      "key": "bogons",
                      "value": {
                        "intValue": "100"
                      }
                    }
                  ]
                }
              ],
              "status": {}
            }
          ]
        }
      ]
    }
  ]
}
